### PR TITLE
Add ability to create gift tokens on demand based upon a share code

### PR DESCRIPTION
**Pull request by JakeChampion**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,6 +15,13 @@ Please install [`next-router`](http://git.svc.ft.com/projects/NEXT/repos/router/
 
 Next article also requires the [libxslt](http://xmlsoft.org/libxslt/) C library - and more specifically its [processing tool](http://xmlsoft.org/XSLT/xsltproc2.html). See the [XSLT section](#xslt) below for more information.
 
+Next article also requires a PEM-formatted private key for signing purposes. This private key is not included in the repository, instead the path to where the private key is to be located in the file system is stored within the `.env` file at the root of this repo. This file is ignored from git. The key Next-Article looks for is `PRIVATE_KEY_PATH`.
```
Comment (matthew-andrews) [inline README.md]: could we put this in config vars?

File: server/app.js [hunk 2]
```diff
@@ -1,11 +1,13 @@
 'use strict';
 
+require('dotenv').load();
```
Comment (matthew-andrews) [inline server/app.js]: this happens by `nbt`

**Comment by matthew-andrews** [comment]
ya'll gone need at least a basic test

File: server/app.js [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 var podcastGuidRegex = '[a-z0-9]{24}';
 var articleUuidRegex = '[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}';
 
 var express = require('ft-next-express');
 var bodyParser = require('body-parser');
 var checks = require('./checks/main.js');
+var cookieParser = require('cookie-parser');
```
Comment (matthew-andrews) [inline server/app.js]: our routing layer strips out cookies

File: package.json [hunk 4]
```diff
@@ -21,7 +23,11 @@
     "html-entities": "~1.1.3",
     "lodash": "~3.9.3",
     "n-podcast-mapping": "Financial-Times/n-podcast-mapping",
-    "next-ft-api-client": "^3.9.5"
+    "next-ft-api-client": "^3.9.5",
+    "node-rsa": "^0.2.25",
+    "share-code-creator": "ssh://git@git.svc.ft.com:7999/labsprot/share-code-generator.git",
```
Comment (matthew-andrews) [inline package.json]: this aint gonna work :wink:

File: client/main.js [hunk 5]
```diff
@@ -69,5 +71,8 @@ require('next-js-setup').bootstrap(function(result) {
 
 	window.addEventListener('load', function() {
 		comments.init(uuid, flags);
+		var user = localStorage['o-tracking-proper-id'];
```
Comment (matthew-andrews) [inline client/main.js]: we use `superstore` for localstorage

File: package.json [hunk 6]
```diff
@@ -21,7 +24,11 @@
     "html-entities": "~1.1.3",
     "lodash": "~3.9.3",
     "n-podcast-mapping": "Financial-Times/n-podcast-mapping",
-    "next-ft-api-client": "^3.9.5"
+    "next-ft-api-client": "^3.9.5",
+    "node-rsa": "^0.2.25",
+    "share-code-creator": "git+ssh://git@git.svc.ft.com:7999/labsprot/share-code-generator.git",
```
Comment (matthew-andrews) [inline package.json]: this is not going to work

Comment (matthew-andrews) [inline package.json]: you can either publish that code publicly somewhere, or we can move `next-article` to stash and deploy with jenkins.

File: package.json [hunk 7]
```diff
@@ -2,11 +2,14 @@
   "name": "ft-next-article",
   "private": true,
   "description": "Render a Next FT Article",
-  "main": "app.js",
+  "main": "server/app.js",
   "scripts": {
     "test": "make test -j 3"
   },
   "dependencies": {
+    "article-token-creator": "git+ssh://git@git.svc.ft.com:7999/labsprot/url-sharing-token-generator.git",
```
Comment (matthew-andrews) [inline package.json]: same with this… we don't / shouldn't have access into the firewall from circle.

File: package.json [hunk 8]
```diff
@@ -2,11 +2,14 @@
...
+    "base-64": "^0.1.0",
+    "base64-url": "^1.2.1",
```
Comment (matthew-andrews) [inline package.json]: these two base64 deps seem unused

File: package.json [hunk 9]
```diff
@@ -21,7 +24,11 @@
     "html-entities": "~1.1.3",
     "lodash": "~3.9.3",
     "n-podcast-mapping": "Financial-Times/n-podcast-mapping",
-    "next-ft-api-client": "^3.9.5"
+    "next-ft-api-client": "^3.9.5",
+    "node-rsa": "^0.2.25",
+    "share-code-creator": "git+ssh://git@git.svc.ft.com:7999/labsprot/share-code-generator.git",
+    "text-encoding": "^0.5.2",
```
Comment (matthew-andrews) [inline package.json]: seems unused

File: package.json [hunk 10]
```diff
@@ -21,7 +24,11 @@
...
+    "uuid": "^2.0.1"
```
Comment (matthew-andrews) [inline package.json]: seems unused

**Comment by JakeChampion** [comment]
New PR https://github.com/Financial-Times/next-article/pull/976

**Pull request closed without merging**